app/validators/date_validator.rb

```diff
@@ -1,9 +1,36 @@
 class DateValidator < ActiveModel::EachValidator
+  RESTRICTION_TYPES = {
+    after: :>,
+    before: :<,
+    on_or_after: :>=,
+    on_or_before: :<=,
+  }.freeze
+
+  DEFAULT_CHECK_VALUES = {
+    today: Date.current,
+    now: Time.current,
+  }.freeze
+
   def validate_each(record, attribute, value)
-    return if value.nil?
+    return record.errors.add(attribute, :invalid) if value.blank? || value.is_a?(Hash)
+
+    restrictions = RESTRICTION_TYPES.keys & options.keys
+
+    restrictions.each do |restriction|
+      operator = RESTRICTION_TYPES[restriction]
+      restriction_option = options[restriction]
+
+      if DEFAULT_CHECK_VALUES.key?(restriction_option)
+        value_to_compare = DEFAULT_CHECK_VALUES[restriction_option]
+      elsif record.respond_to?(restriction_option)
+        value_to_compare = record.send(restriction_option)
+      end
+
+      next if value_to_compare.blank? || value_to_compare.is_a?(Hash)
+
+      raise ArgumentError, "give me something to work with!!" unless value_to_compare
```

Comment:
If value_to_compare is blank, this line will never be read, because `next` would have been triggered on line 29, right?

Reply:
Correct you are, good spot. 

Theres a conflict here between when a date is nil because it couldn't be cast as a date (e.g `validates :starts_on, date: { after: :expires_at }` when `expires_at` is invalid/incomplete/blank which means we don't want to do the validation, and when we pass in something that isn't valid (i.e it isn't in `DEFAULT_CHECK_VALUES` or the record doesn't `respond_to` it which means we want to raise an error.

Reply:
*updated

what do you think of that?